### PR TITLE
KTOR-3944 SinglePageApplication plugin returns 404 for non-existent paths

### PR DESCRIPTION
**Pull request by lolilofit**
[KTOR-3944](https://youtrack.jetbrains.com/issue/KTOR-3944)

### CODE REVIEW

File: ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/SinglePageApplication.kt [hunk 1]
```diff
@@ -26,45 +21,34 @@ import java.io.*
...
-
-        if (ignoredFiles.firstOrNull { it.invoke(requestUrl) } != null) {
-            call.respond(HttpStatusCode.Forbidden)
+            if (config.useResources) {
+                resourceWithDefault(
+                    config.filesPath,
+                    config.defaultPage,
+                    shouldFileBeIgnored
+                )
+                defaultResource(config.defaultPage, config.filesPath)
```
Comment (ishitatsuyuki) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/SinglePageApplication.kt:40]: Is a default fallback needed in presence of resource/filesWithDefault? Routes ending with a slash should be handled by the "not found" case.

Comment (lolilofit) [reply ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/SinglePageApplication.kt:40]: Thanks, you are right, I'will remove this line 👍

**Review by e5l — APPROVED** [review]
@andreyaksenov could you check docs?

File: ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/SinglePageApplication.kt [hunk 2]
```diff
@@ -2,69 +2,53 @@
...
  * ```
- * install(SinglePageApplication) {
- *   filesPath = "application/project_path"
+ * application {
+ *     routing {
+ *        singlePageApplication {
+ *           filesPath = "application/project_path"
+ *         }
+ *     }
  * }
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/SinglePageApplication.kt:25]: Let's add a link to docs:

> You can learn more from [Serving single-page applications](https://ktor.io/docs/serving-spa.html).

Note that this link will return 404 until 2.0.0 is released.

File: ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/SinglePageApplication.kt [hunk 3]
```diff
@@ -2,69 +2,53 @@
...
 import io.ktor.server.request.*
-import io.ktor.server.response.*
 import io.ktor.server.routing.*
-import io.ktor.util.*
-import io.ktor.util.pipeline.*
 import java.io.*
 
 /**
- * A plugin that allows you to serve a single-page application
+ * Serves a single-page application
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/SinglePageApplication.kt:13]: > Serves a single-page application.

Add a dot

File: ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/SinglePageApplication.kt [hunk 4]
```diff
@@ -2,69 +2,53 @@
...
  *
- * A basic plugin configuration for the application served from the filesPath folder
+ * A basic configuration for the application served from the filesPath folder
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/SinglePageApplication.kt:15]: > from the `filesPath` folder

File: ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/SinglePageApplication.kt [hunk 5]
```diff
@@ -2,69 +2,53 @@
...
  * with index.html as a default file:
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/SinglePageApplication.kt:16]: > with `index.html` as a default file:

File: ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/SinglePageApplication.kt [hunk 6]
```diff
@@ -2,69 +2,53 @@
...
+        } else {
+            staticRootFolder = File(config.filesPath)
+            filesWithDefault(".", config.defaultPage, shouldFileBeIgnored)
         }
     }
 }
 
 /**
- * Configuration for the [SinglePageApplication] plugin
+ * Configuration for the [Route.singlePageApplication]
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/SinglePageApplication.kt:51]: > ... for the [Route.singlePageApplication] plugin.

File: ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt [hunk 7]
```diff
@@ -125,6 +125,18 @@ public fun Route.file(remotePath: String, localPath: File) {
  */
 public fun Route.files(folder: String): Unit = files(File(folder))
 
+/**
+ * Sets up routing to serve all files from [folder]
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt:129]: > ... from [folder].

dot

**Review by andreyaksenov — CHANGES_REQUESTED** [review]

File: ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt [hunk 8]
```diff
@@ -125,6 +125,18 @@ public fun Route.file(remotePath: String, localPath: File) {
  */
 public fun Route.files(folder: String): Unit = files(File(folder))
 
+/**
+ * Sets up routing to serve all files from [folder].
+ * Serves [defaultFile] if not existing file is requested.
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt:130]: > Serves [defaultFile] if a missing file is requested.

File: ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt [hunk 9]
```diff
@@ -125,6 +125,18 @@ public fun Route.file(remotePath: String, localPath: File) {
...
+ * Serves [defaultFile] if the requested file should be ignored by [shouldFileBeIgnored]
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt:131]: please add dot at the end

File: ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt [hunk 10]
```diff
@@ -138,6 +150,35 @@ public fun Route.files(folder: File) {
     }
 }
 
+/**
+ * Sets up routing to serve all files from [folder].
+ * Serves [defaultFile] if not existing file is requested.
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt:155]: > Serves [defaultFile] if a missing file is requested.

File: ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt [hunk 11]
```diff
@@ -138,6 +150,35 @@ public fun Route.files(folder: File) {
...
+ * Serves [defaultFile] if the requested file should be ignored by [shouldFileBeIgnored]
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt:156]: please add dot at the end

File: ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt [hunk 12]
```diff
@@ -220,6 +261,37 @@ public fun Route.resource(remotePath: String, resource: String = remotePath, res
     }
 }
 
+/**
+ * Sets up routing to serve all resources in [resourcePackage].
+ * Serves [defaultFile] if not existing file is requested.
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt:266]: > Serves [defaultFile] if a missing file is requested.

File: ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt [hunk 13]
```diff
@@ -220,6 +261,37 @@ public fun Route.resource(remotePath: String, resource: String = remotePath, res
...
+ * Serves [defaultFile] if the requested file should be ignored by [shouldFileBeIgnored]
```
Comment (andreyaksenov) [inline ktor-server/ktor-server-core/jvm/src/io/ktor/server/http/content/StaticContent.kt:267]: dot at the end

**Review by andreyaksenov — CHANGES_REQUESTED** [review]
@lolilofit several minor comments

**Review by andreyaksenov — APPROVED** [review]

**Pull request merged**